feat: implement observability of notifications subsystem

dannykopping opened the pull request:
Closes https://github.com/coder/team-coconut/issues/3

Also performed some drive-by refactorings.

_NOTE: this is stacked on `dk/system-notifications-lib, but I didn't want to use Graphite for fear it screws up that branch which is close to being merged; I'm still pretty new to that tool._

dannykopping committed:
Implement observability of notification subsystem

Minor refactoring to make testing easier

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make lint

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make gen

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make fmt

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Small fixes

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -195,9 +200,18 @@ func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotification
+
+	start := time.Now()
 	retryable, err := deliver(ctx, msg.ID)
+	n.metrics.DispatcherSendSeconds.WithLabelValues(string(n.method)).Observe(time.Since(start).Seconds())
```
Reviewer: we may want to distinguish successful vs failed dispatch speed; what do you think?
I left it out because it's probably not very useful in practice.

mtojek replied:
IMHO Dispatching time is fine to be left as is, as a single timer. I would rather consider including the failure reason if possible.

mafredri commented on coderd/database/migrations/000223_notifications_metrics.up.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE notification_messages
+ADD COLUMN queued_seconds FLOAT NULL;
```
Nit: The name `queued_seconds` implies an integer, consider changing the type or the name.

dannykopping replied:
I don't think it necessarily implies integer; `time.Duration`'s [`Seconds()`](https://pkg.go.dev/time#Duration.Seconds) returns a float64.

mafredri replied:
Fair enough 😄.

In the DB, usually it means seconds:

```
    lifetime_seconds bigint DEFAULT 86400 NOT NULL,
    timeout_seconds integer NOT NULL
    connection_timeout_seconds integer DEFAULT 0 NOT NULL,
```

But fine to keep as float if we care about the precision.

dannykopping replied:
I suppose I don't strictly _need_ to keep sub-second precision on the length of time messages were queued for.
I'll switch to `integer` 👍

spikecurtis replied:
This is backing a Prometheus metric, which are always floats, and by convention end in the SI unit, which for time is seconds.  It implies no precision.

This should be a float: there just isn't any compelling reason to pre-judge the required precision. This is not a banking application, nor do we need to worry about the relative speed of integer vs floating point arithmetic.

The name `queued_seconds` is appropriate.

mafredri commented on coderd/notifications/manager.go:
```diff
@@ -56,7 +59,13 @@ type Manager struct {
 // access URL etc.
-func NewManager(cfg codersdk.NotificationsConfig, store Store, log slog.Logger) (*Manager, error) {
+func NewManager(cfg codersdk.NotificationsConfig, store Store, metrics *Metrics, log slog.Logger) (*Manager, error) {
+	// TODO(dannyk): add the ability to use multiple notification methods.
```
Will this be fixed in this PR, a follow-up, or should we track this via issue?

dannykopping replied:
In a follow-up, but the work is not yet planned.
This will need to be implemented once we expand beyond system-only notifications.
I think it's ok to leave untracked for now.

mafredri replied:
The reason I lifted this up is that `// TODO`s in code is usually where TODO's go to die, which is why I prefer them either tracked or not important enough to mention. Not a blocker to leave it in, just a pet-peeve.

Exhibit A:

```
rg 'TODO\(mafredri\)' | wc -l
      31
```

mtojek commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -929,15 +929,20 @@ func (q *FakeQuerier) AcquireNotificationMessages(_ context.Context, arg databas
 	}
 
+	// Shift the first "Count" notifications off the slice (FIFO).
+	sz := len(q.notificationMessages)
```
why `sz` :) ?

dannykopping replied:
`size`

dannykopping replied:
I generally agree, although I try to follow this advice: https://go.dev/talks/2014/names.slide#4

> The greater the distance between a name's declaration and its uses, the longer the name should be.

Here it's pretty clear what `sz` does, so brevity is helpful.

mtojek replied:
offtopic: It depends on your background. For me `sz` is reserved for **s**tring **z**ero-terminated (c++).

dannykopping replied:
Ha! Fair enough :)

mafredri commented on coderd/notifications/manager.go:
```diff
@@ -56,7 +59,13 @@ type Manager struct {
+	// TODO(dannyk): add the ability to use multiple notification methods.
+	var method database.NotificationMethod
+	if err := method.Scan(cfg.Method.String()); err != nil {
+		return nil, xerrors.Errorf("given notification method %q is invalid", cfg.Method)
```
Nit: simplify (given is given 😄?)
```suggestion
		return nil, xerrors.Errorf("notification method %q is invalid", cfg.Method)
```

mtojek commented on coderd/database/migrations/000223_notifications_metrics.up.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE notification_messages
+ADD COLUMN queued_seconds FLOAT NULL;
```
Shouldn't we go with `numeric`?

dannykopping replied:
It's been a while since I've used PG so I had to look up the difference.

Seems like `FLOAT` might be more appropriate here since we don't need the absolute precision that `NUMERIC` provides and `FLOAT` is faster/smaller.

Is there some other reason I should consider using `NUMERIC` over `FLOAT` which I might be missing?

mtojek replied:
Passing the database question to @mafredri. My understanding is that floats can always lose precision, so it is safer to store round numbers. If we need fractions of seconds, let's go with milliseconds?

mafredri replied:
I prefer ints as well, but not enough to block this PR, so I'll leave it up to Danny.

dannykopping replied:
As discussed in https://github.com/coder/coder/pull/13799#discussion_r1668521931, I actually don't specifically need sub-second precision so I'll be changing over to `integer`.

mafredri commented on coderd/notifications/metrics_test.go:
```diff
@@ -0,0 +1,342 @@
+	}
+
+	// then
+	require.Eventually(t, func() bool {
```
Note: Polling designs are often flakey, it would be nice if we could refactor it. Something to keep in mind for the future.

dannykopping replied:
Indeed, I have https://github.com/coder/team-coconut/issues/6 scheduled.

mafredri commented on coderd/notifications/notifier.go:
```diff
@@ -195,9 +200,18 @@ func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotification
-	logger := n.log.With(slog.F("msg_id", msg.ID), slog.F("method", msg.Method))
+	logger := n.log.With(slog.F("msg_id", msg.ID), slog.F("method", msg.Method), slog.F("attempt", msg.AttemptCount+1))
+
+	if msg.AttemptCount > 0 {
```
I take it this is incremented somewhere not visible in this PR?

dannykopping replied:
Correct; in the database layer.

mafredri reviewed (approved):
FWIW the changes LGTM in isolation (I haven't reviewed the original code).

mtojek commented on coderd/notifications/metrics_test.go:
```diff
@@ -0,0 +1,342 @@
+	t.Parallel()
+
+	// setup
+	if !dbtestutil.WillUsePostgres() {
```
Is it because of pubsub or something else?

dannykopping replied:
A lot of the business logic is implemented in the database layer.
I'm requiring postgres here because the full functionality is required (retries, etc).
I'll clarify this inline.

mtojek commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+		}, []string{LabelMethod}),
+
+		// Aggregating on LabelTemplateID as well would cause a cardinality explosion.
+		DispatcherSendSeconds: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
```
DispatchingTimeSeconds :) ?

dannykopping replied:
Seems a bit redundant to have "time" and "seconds" in the name.
I think this is already clear enough, TBH.

mtojek reviewed (approved):
Nothing blocking 👍

dannykopping committed:
Apply suggestions from code review

Co-authored-by: Mathias Fredriksson <mafredri@gmail.com>

mtojek reviewed (approved):
Thanks for all replies.

dannykopping committed:
Review comments

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Correcting query

Signed-off-by: Danny Kopping <danny@coder.com>

spikecurtis commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+		PermFailureCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "permanent_failures_count", Namespace: ns, Subsystem: subsystem,
+			Help: "The count of notifications which failed and have exceeded their retry attempts.",
+		}, []string{LabelMethod, LabelTemplateID}),
```
DispatchCount, TempFailureCount, and PermFailureCount might make sense as a single metric with different "disposition" (or "result") labels.  The help text refers to them as "notifications", but I think that's misleading.  What we are counting is "notification attempts" or "dispatch attempts", which might individually end up in success or some kind of failure.  In particular, in the case of failure and retry, the number of delivery attempts can exceed the number of notifications, so it's important to make the distinction.

Prometheus guidelines say that something should be a single metric with labels (vs multiple metrics) when the sum or average of the label is sensible and coherent. Filtering by method and summing across disposition would be useful, for example, to track the number of web requests we are making via Webhook delivery.

spikecurtis commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+		PendingUpdates: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "pending_updates", Namespace: ns, Subsystem: subsystem,
+			Help: "The number of updates waiting to be flushed to the store.",
+		}),
```
A really useful Gauge is the number of "in progress" delivery attempts.  You increment it when you kick off the delivery goroutine and decrement it when you send to `success` or `failure` channels.  Can help spot hung handlers/leaked goroutines

spikecurtis replied:
Also a count of success/failure updates flushed to the database.

spikecurtis commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+func NewMetrics(reg prometheus.Registerer) *Metrics {
+	return &Metrics{
+		DispatchedCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "dispatched_count", Namespace: ns, Subsystem: subsystem,
```
accumulating counts should end in the units and _total_ as in `coderd_notifications_successful_deliveries_total`, where "deliveries" is the unit (could use "dispatch_attempts", or something similar).

https://prometheus.io/docs/practices/naming/

spikecurtis commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+		// Currently no requirement to discriminate between success and failure updates which are pending.
+		PendingUpdates: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "pending_updates", Namespace: ns, Subsystem: subsystem,
+			Help: "The number of updates waiting to be flushed to the store.",
```
Sounds too generic to me.  "The number of delivery attempt results waiting to be flushed to the store" ?

spikecurtis commented on coderd/database/migrations/000223_notifications_metrics.up.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE notification_messages
+ADD COLUMN queued_seconds INTEGER NULL;
```
Do we really to store this on the row?  It feels a little strange and ambiguous.  If the notification_message is currently queued, then it's a bit meaningless, since we don't know for how much longer it will be queued. I.e. it is at present being returned on Enqueue, but is meaningless in that context.

When we dequeue it, can't the notifier compute how long it was on the queue based on it's `created_at` or `updated_at` time?  Or, if we really want to ensure all the measurements are based on the central PostgreSQL server clock, then it can part of the result returned to the notifier without having to write it back to the database.

spikecurtis commented on coderd/notifications/notifier.go:
```diff
@@ -226,13 +240,44 @@ func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotification
+			success <- n.newSuccessfulDispatch(msg)
 		}
 	}
+	n.metrics.PendingUpdates.Set(float64(len(success) + len(failure)))
```
I think it would be cleaner to just compute this metric on-`Collect`, rather than `Set`ing it when something reads or writes to these channels.  As an illustration of how easy it is to miss a spot, note that you have forgotten to set it in `process()` after a failed call to `prepare()`.

spikecurtis commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+		// Aggregating on LabelTemplateID as well would cause a cardinality explosion.
+		QueuedSeconds: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name: "queued_seconds", Namespace: ns, Subsystem: subsystem,
+			Buckets: []float64{1, 5, 15, 30, 60, 120, 300, 600, 3600, 86400},
```
We say this should be less than the fetch interval, which is 15s --- so I think there should be more buckets <15s.  In fact, it should be more like 15/(n+1) where n is the number of Coderd replicas, assuming that notifications usually get picked up by the next fetch.  Seeing the average move up toward 15s is way more important than the distinction between 10minutes, 1 hour and 1 day, all of which indicate that notifications are just blocked entirely.

spikecurtis commented on coderd/notifications/metrics.go:
```diff
@@ -0,0 +1,71 @@
+		// Aggregating on LabelTemplateID as well would cause a cardinality explosion.
+		DispatcherSendSeconds: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name: "dispatcher_send_seconds", Namespace: ns, Subsystem: subsystem,
+			Buckets: []float64{0.001, 0.05, 0.1, 0.5, 1, 2, 5, 10, 15, 30, 60, 120},
```
should be `0.001, 0.005, 0.01, 0.05` on the bottom end, right?

spikecurtis replied:
Dispatching, say, a webhook, within the same cluster could be a few milliseconds latency.  Same geographic region could be 10-20ms.

Almost nothing is going to be 1ms unless it's extremely simple and extremely close, but in 50ms you could be dispatching across an entire continent.  With buckets like this you could have something that normally returns in 15ms take 3x its normal time and be close to falling over and this metric would not clock it.  That seems much more important than 30s, 60s, 120s.

dannykopping replied:
I hear you and I can see where you're coming from.

Let's wait for users to ask for this. I think there's sufficient coverage for now, and the additional datapoints are likely not going to be very useful for the majority of operators IMHO.

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into dk/system-notifications-o11y

dannykopping committed:
Only return UUID from EnqueueNotificationMessage

Changes made to dbgen are to avoid this panic:

$ ./coderd/database/generate.sh
generate
panic: unknown return type: *dst.SelectorExpr

goroutine 1 [running]:
main.orderAndStubDatabaseFunctions({0x1400001ea80, 0x3c}, {0x102d019e0, 0x1}, {0x102d0bc75, 0xb}, 0x102df7b20)
        /Users/danny/Code/coder/coder/scripts/dbgen/main.go:367 +0x2588
main.run()
        /Users/danny/Code/coder/coder/scripts/dbgen/main.go:56 +0x118
main.main()
        /Users/danny/Code/coder/coder/scripts/dbgen/main.go:42 +0x1c
exit status 2

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Review feedback

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -215,24 +234,55 @@ func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotification
 			logger.Warn(context.Background(), "cannot record dispatch failure result", slog.Error(ctx.Err()))
 			return ctx.Err()
-		default:
+		case failure <- n.newFailedDispatch(msg, err, retryable):
```
Addresses https://github.com/coder/coder/pull/13537#discussion_r1668844087

dannykopping committed:
Minor fixups

Signed-off-by: Danny Kopping <danny@coder.com>

mafredri commented on coderd/database/queries/notifications.sql:
```diff
@@ -19,7 +19,7 @@ VALUES (@id,
         @targets,
         @created_by)
-RETURNING *;
+RETURNING id;
```
I think you can revert the hack if you change this to `RETURNING id::uuid;`. Alternatively, in `VALUES(`, using `@id::uuid` could also work.

mafredri replied:
Actually, I may be wrong. Not really sure what you ran into 😄

dannykopping replied:
Oh nice! That worked 🥇

mafredri commented on coderd/notifications/metrics_test.go:
```diff
@@ -80,40 +83,44 @@ func TestMetrics(t *testing.T) {
-			if debug {
-				t.Logf("coderd_notifications_dispatched_count == 1: %v", metric.Counter.GetValue())
-			}
+				target = &val
```
Linter unhappy, this should be fine with Go 1.22, though.

mtojek reviewed (approved):
👍

dannykopping committed:
Revert hack, no output param needed

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Small touch-ups

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Merge branch 'main' of https://github.com/coder/coder into dk/system-notifications-o11y

dannykopping committed:
Merge branch 'main' of https://github.com/coder/coder into dk/system-notifications-o11y

dannykopping committed:
Harden tests, fail early

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make fmt

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Restoring deleted line

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented on coderd/coderd.go:
```diff
@@ -1018,7 +1018,6 @@ func New(options *Options) *API {
 					})
 					r.Put("/appearance", api.putUserAppearanceSettings)
 					r.Route("/password", func(r chi.Router) {
-						r.Use(httpmw.RateLimit(options.LoginRateLimit, time.Minute))
```
Whoops!

dannykopping committed:
Comments

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Lock before modification

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Remove TestNotifierPaused's unnecessarily fast fetch interval

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Merge branch 'main' of https://github.com/coder/coder into dk/system-notifications-o11y

dannykopping committed:
Rename migration after numbering conflict

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Small fixes

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Merge branch 'main' of https://github.com/coder/coder into dk/system-notifications-o11y

dannykopping committed:
Logging improvement

Signed-off-by: Danny Kopping <danny@coder.com>

mtojek reviewed (approved):
Stamp 👍
